### PR TITLE
feature/autogenerate_mongo_db_ids-GDEV-159

### PR DESCRIPTION
**Pull request by alamgirmani**

### CODE REVIEW

File: metadata_service/dao/study.py [hunk 1]
```diff
@@ -87,7 +98,11 @@ async def update_study(study_id: str, data: Dict) -> Dict:
...
     """
-    collection = await get_collection(COLLECTION_NAME)
-    await collection.update_one({"id": study_id}, {"$set": data})
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    await collection.update_one(  # type: ignore
+        {"id": study_id}, {"$set": data.dict(exclude_unset=True)}
+    )
     study = await get_study(study_id)
+    await db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/study.py:107]: This is aggressively closing the db connection. The connection is closed by `get_study` in the line above.

File: metadata_service/dao/study.py [hunk 2]
```diff
@@ -69,14 +76,18 @@ async def add_study(data: Dict) -> Dict:
       The added Study object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    study_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    study_id = await db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    print(data)
```
Comment (deepakunni3) [inline metadata_service/dao/study.py:82]: Remove errant print statement

File: metadata_service/dao/publication.py [hunk 3]
```diff
@@ -87,7 +95,9 @@ async def update_publication(publication_id: str, data: Dict) -> Dict:
       The updated Publication object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    await collection.update_one({"id": publication_id}, {"$set": data})
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    await collection.update_one({"id": publication_id}, {"$set": data.dict()})  # type: ignore
     publication = await get_publication(publication_id)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/publication.py:102]: Need to await the async call.

File: metadata_service/dao/publication.py [hunk 4]
```diff
@@ -69,14 +74,17 @@ async def add_publication(data: Dict) -> Dict:
...
-    collection = await get_collection(COLLECTION_NAME)
-    publication_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    publication_id = db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = publication_id
+    await collection.insert_one(data.dict())  # type: ignore
     publication = await get_publication(publication_id)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/publication.py:83]: Need to await the async call.

File: metadata_service/dao/publication.py [hunk 5]
```diff
@@ -69,14 +74,17 @@ async def add_publication(data: Dict) -> Dict:
       The added Publication object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    publication_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    publication_id = db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = publication_id
```
Comment (deepakunni3) [inline metadata_service/dao/publication.py:80]: Cannot mutate a Pydantic model object. Leads to an exception.

File: metadata_service/dao/publication.py [hunk 6]
```diff
@@ -47,19 +50,21 @@ async def get_publication(publication_id: str, embedded: bool = False) -> Dict:
...
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    publication = await collection.find_one({"id": publication_id})  # type: ignore
     if not publication:
         raise HTTPException(
             status_code=404,
             detail=f"{Publication.__name__} with id '{publication_id}' not found",
         )
     if embedded:
         publication = await embed_references(publication, Publication)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/publication.py:63]: Need to await the async call.

File: metadata_service/dao/publication.py [hunk 7]
```diff
@@ -14,29 +14,32 @@
...
     Returns:
       A list of Publication objects.
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    publications = await collection.find().to_list(None)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    publications = await collection.find().to_list(None)  # type: ignore
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/publication.py:38]: Need to await the async call.

File: metadata_service/dao/file.py [hunk 8]
```diff
@@ -86,7 +92,9 @@ async def update_file(file_id: str, data: Dict) -> Dict:
       The updated File object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    await collection.update_one({"id": file_id}, {"$set": data})
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    await collection.update_one({"id": file_id}, {"$set": data.dict()})  # type: ignore
     file = await get_file(file_id)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/file.py:99]: Need to await the async call.

File: metadata_service/dao/file.py [hunk 9]
```diff
@@ -68,14 +72,16 @@ async def add_file(data: Dict) -> Dict:
...
     """
-    collection = await get_collection(COLLECTION_NAME)
-    file_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    file_id = data["id"]  # type: ignore
+    await collection.insert_one(data.dict())  # type: ignore
     file = await get_file(file_id)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/file.py:80]: Need to await for async call.

File: metadata_service/dao/file.py [hunk 10]
```diff
@@ -14,29 +14,31 @@
...
     Returns:
       A list of File objects.
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    files = await collection.find().to_list(None)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    files = await collection.find().to_list(None)  # type: ignore
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/file.py:37]: Need to await the async call.

File: metadata_service/dao/experiment.py [hunk 11]
```diff
@@ -87,7 +95,9 @@ async def update_experiment(experiment_id: str, data: Dict) -> Dict:
       The updated Experiment object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    await collection.update_one({"id": experiment_id}, {"$set": data})
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    await collection.update_one({"id": experiment_id}, {"$set": data.dict()})  # type: ignore
     experiment = await get_experiment(experiment_id)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/experiment.py:102]: Need to await the async call.

File: metadata_service/dao/experiment.py [hunk 12]
```diff
@@ -69,14 +74,17 @@ async def add_experiment(data: Dict) -> Dict:
...
-    collection = await get_collection(COLLECTION_NAME)
-    experiment_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    experiment_id = await db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = experiment_id
+    await collection.insert_one(data.dict())  # type: ignore
     experiment = await get_experiment(experiment_id)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/experiment.py:83]: Await.

File: metadata_service/dao/experiment.py [hunk 13]
```diff
@@ -14,29 +14,32 @@
...
     Returns:
       A list of Experiment objects.
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    experiments = await collection.find().to_list(None)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    experiments = await collection.find().to_list(None)  # type: ignore
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/experiment.py:38]: Await.

File: metadata_service/dao/experiment.py [hunk 14]
```diff
@@ -69,14 +74,17 @@ async def add_experiment(data: Dict) -> Dict:
       The added Experiment object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    experiment_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    experiment_id = await db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = experiment_id
```
Comment (deepakunni3) [inline metadata_service/dao/experiment.py:80]: Cannot mutate a Pydantic model object. Leads to an exception.

File: metadata_service/dao/data_access_committee.py [hunk 15]
```diff
@@ -70,14 +75,17 @@ async def add_dac(data: Dict) -> Dict:
       The added DAC object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    dac_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    dac_id = await db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = dac_id
```
Comment (deepakunni3) [inline metadata_service/dao/data_access_committee.py:81]: Cannot mutate a Pydantic model object. Leads to an exception.

File: metadata_service/dao/data_access_policy.py [hunk 16]
```diff
@@ -69,24 +74,32 @@ async def add_dap(data: Dict) -> Dict:
       The added DAP object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    dap_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    dap_id = db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = dap_id
```
Comment (deepakunni3) [inline metadata_service/dao/data_access_policy.py:80]: Cannot mutate a Pydantic model object. Leads to an exception.

File: metadata_service/dao/dataset.py [hunk 17]
```diff
@@ -70,14 +75,17 @@ async def add_dataset(data: Dict) -> Dict:
       The added Dataset object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    dataset_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    dataset_id = await db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = dataset_id
```
Comment (deepakunni3) [inline metadata_service/dao/dataset.py:81]: Cannot mutate a Pydantic model object. Leads to an exception.

File: metadata_service/dao/data_access_policy.py [hunk 18]
```diff
@@ -69,24 +74,32 @@ async def add_dap(data: Dict) -> Dict:
...
+    await collection.insert_one(data.dict())  # type: ignore
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/data_access_policy.py:82]: Await.

File: metadata_service/dao/dataset.py [hunk 19]
```diff
@@ -15,29 +15,32 @@
...
     Returns:
       A list of Dataset objects.
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    datasets = await collection.find().to_list(None)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    datasets = await collection.find().to_list(None)  # type: ignore
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/dataset.py:39]: Await.

File: metadata_service/dao/dataset.py [hunk 20]
```diff
@@ -48,19 +51,21 @@ async def get_dataset(dataset_id: str, embedded=False) -> Dict:
...
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    dataset = await collection.find_one({"id": dataset_id})  # type: ignore
     if not dataset:
         raise HTTPException(
             status_code=404,
             detail=f"{Dataset.__name__} with id '{dataset_id}' not found",
         )
     if embedded:
         dataset = await embed_references(dataset, Dataset)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/dataset.py:64]: Await.

File: metadata_service/dao/dataset.py [hunk 21]
```diff
@@ -70,14 +75,17 @@ async def add_dataset(data: Dict) -> Dict:
...
     """
-    collection = await get_collection(COLLECTION_NAME)
-    dataset_id = data["id"]
-    await collection.insert_one(data)
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    dataset_id = await db_connect.get_next_id(COLLECTION_NAME, PREFIX)
+    data["id"] = dataset_id
+    await collection.insert_one(data.dict())  # type: ignore
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/dataset.py:83]: Await.

File: metadata_service/dao/dataset.py [hunk 22]
```diff
@@ -88,7 +96,9 @@ async def update_dataset(dataset_id: str, data: Dict) -> Dict:
       The updated Dataset object
 
     """
-    collection = await get_collection(COLLECTION_NAME)
-    await collection.update_one({"id": dataset_id}, {"$set": data})
+    db_connect = DBConnect()
+    collection = await db_connect.get_collection(COLLECTION_NAME)
+    await collection.update_one({"id": dataset_id}, {"$set": data.dict()})  # type: ignore
     dataset = await get_dataset(dataset_id)
+    db_connect.close_db()
```
Comment (deepakunni3) [inline metadata_service/dao/dataset.py:103]: No need to close the db connection. This is already being taken care of by the `get_dataset` method in the line above.

**Review by deepakunni3 — CHANGES_REQUESTED** [review]

File: metadata_service/dao/study.py (hunk 1 above)
Comment (alamgirmani) [reply metadata_service/dao/study.py:105]: fixed it

File: metadata_service/dao/study.py (hunk 2 above)
Comment (alamgirmani) [reply metadata_service/dao/study.py]: removed

File: metadata_service/dao/publication.py (hunk 5 above)
Comment (alamgirmani) [reply metadata_service/dao/publication.py]: changed it to data.id

File: metadata_service/dao/dataset.py (hunk 20 above)
Comment (alamgirmani) [reply metadata_service/dao/dataset.py]: await added

File: metadata_service/dao/dataset.py (hunk 21 above)
Comment (alamgirmani) [reply metadata_service/dao/dataset.py]: await added

File: metadata_service/dao/dataset.py (hunk 22 above)
Comment (alamgirmani) [reply metadata_service/dao/dataset.py]: moved the close method one line above so as to match the get and close connection

**Review by deepakunni3 — APPROVED** [review]

**Pull request merged**